### ISSUE TITLE
Allow podman run against specgen json

### ISSUE THREAD

**Issue by coandco** [containers/podman#14264]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

Now that specgen has been merged into mainline podman, it would be nice to give `podman run` the ability to load a generated spec json from disk, rather than needing to pass dozens of options to it.

Specifically, this would solve some issues that neither the docker-compose shim nor `podman play kube` solve, because it would allow control of podman-specific knobs like `--cgroups=split` and `--sdnotify=conmon`.

**Comment by mheon** [comment]
I'm in favor of this - seems easy to implement and useful.

**Comment by atlantis** [comment]
This would be super useful for us - it would allow an ecosystem of tools to be built which manipulate the JSON, which is much cleaner than cobbling together and then saving a string of complex command line params.

**Comment by vrothberg** [comment]
SGTM, a really cool idea

Cc: @baude @rhatdan

**Comment by vrothberg** [comment]
> How should this work with other options? Should it
> a) allow to set other options to overwrite stuff from the given specgen or

That sounds reasonable to me.

The user-specified specgen would be used for "initialization" and options can override.  In a way, that can be an alternative to using custom containers.conf, so we could use an environment variable to specify that specgen in addition to a CLI flag.

**Comment by mheon** [comment]
I almost see this as a separate command, with no options beyond the Specgen to use - seems like that's the easiest way to implement. Accepting a base SpecGen in run could be complicated in that we'd have to deal with defaults overwriting it.

**Comment by rhatdan** [comment]
podman play specgen ...?

**Comment by vrothberg** [comment]
> podman play specgen ...?

There was a plan to have `podman kube {down,play}` etc. so I am no fan of adding more things to `play`. The desired command really belongs under `podman container`.

I imagine rather a `--specgen=path` flag for `podman {create,run}`.  If it's a `SpecGenerator`, the code changes would be minimal.

**Comment by rhatdan** [comment]
My problem with it being a part of a current command is that you need to say all other options are not available.
podman run --specgen=path --security-opt seccomp:unconfined -d sleep 1000
podman run --specgen=path --security-opt seccomp:unconfined -d alpine sleep 1000

What do we do here?

**Comment by rhatdan** [comment]
How about 
podman container createspec SPEC.FILE
podman container runspec SPEC.FILE

**Comment by vrothberg** [comment]
> My problem with it being a part of a current command is that you need to say all other options are not available. 

Can you elaborate? I think they can still be used, see below.

> podman run --specgen=path --security-opt seccomp:unconfined -d sleep 1000 podman run --specgen=path --security-opt seccomp:unconfined -d alpine sleep 1000
> 
> What do we do here?

The specified spec is the baseline.  Other CLI options allow for changing that baseline.

If the spec says "option=foo" and we have `--option=bar`, the container will run with "bar".

But one thing that just came to my mind is how to deal with the `IMAGE [ARGS]`.  Since the specgen could have both specified, it remains largely unclear whether the first unknown argument would refer to an image or a command.

> How about
podman container createspec SPEC.FILE
podman container runspec SPEC.FILE

SGTM :heavy_check_mark:

**Comment by cdoern** [comment]
I like this idea... I can implement this summer if no one else wants to take this?

**Comment by cdoern** [comment]
working on this...

**Comment by baude** [comment]
@coandco as the original reporter, i'm trying to understand why you needed this?  can you elaborate on your use case? more verbose == better so I can understand it more.

**Comment by rhatdan** [comment]
Makes sense to me, I would like to make it a bit more human usable though. YAML rather the json for example.

**Comment by coandco** [comment]
Well, since JSON is a strict subset of YAML, if you have it parse the file as YAML it'll transparently work with autogenerated JSON specs while still allowing users to make their own YAML specs, yes?

**Comment by rhatdan** [comment]
My point being that humans find it easier in general to work with YAML files than JSON files.  If we are going to export the SPEC, I would prefer that the exported file be a YAML file, so that you could easily edit it and create a new container off of it.  As opposed to handling you a JSON file that you need to edit.

**Comment by coandco** [comment]
Sure.  Either way, going from "no way to specify all of a container's options in a file" to "able to specify all of a container's options in a file" feels like a huge improvement.  "The specfile is in JSON" to "the specfile is in YAML" feels like a modest quality of life improvement.

**Comment by vrothberg** [comment]
> I've looked into Quadlet, and several aspects of the above example systemd unit file are based on the best practices found in https://github.com/containers/quadlet/blob/main/docs/ContainerSetup.md, but I'd like to retain the ability to tweak individual aspects of the container setup to my needs. I believe that having a docker-compose-like ability to fully specify a container from a specfile would neatly solve my problem.

I have a "feeling" that extending Quadlet for your needs may be worth a shot.  There are plans to move Quadlet over into this repository (and migrate it to Go), such that Podman and Quadlet can share the generator code.

**Comment by rhatdan** [comment]
I like extending Quadlet as much as possible, but eventually it is going to get very complex and we will have to add a get out of jail free card to it to allow the specificaion of addiitonal_podman_options flag to the unit file.

I think we could just add an experimental feature for now to allow the use of the specgen.

**Comment by rhatdan** [comment]
BTW I was talking to @mattdm yesterday, and it pointed out that as an administrator he loves the idea of Quadlet, since it allows him to specify what he wants for running a container in a unit file, and then as Podman improves, his service gets the improvement without him needed to change anything.  
IE If we improve the general security of podman or take advantage of new features of the OS, he can get them as default.  Whereas with the current podman generate systemd, he thinks that we hard code some flags of Podman into the unit file.

Specifically he has watched as podman generate systemd has evolved. Just food for thought.

**Comment by vrothberg** [comment]
I recall the mail thread :) I chatted with @baude to prioritize migrating Quadlet over to containers/podman. This way, Quadlet and Podman are always in sync and share the generator code.

**Comment by vrothberg** [comment]
Yes, I am in favor of a dedicated meeting for this topic.

**Comment by coandco** [comment]
...wouldn't it make sense to implement run-from-spec even if Quadlet is the way forward for systemd units?  It seems like run-from-spec is pretty low-lift, high-reward, since the groundwork for it has already been laid.

**Comment by vrothberg** [comment]
Apologies, I didn't mean to but clicked on the wrong button.

> ...wouldn't it make sense to implement run-from-spec even if Quadlet is the way forward for systemd units? It seems like run-from-spec is pretty low-lift, high-reward, since the groundwork for it has already been laid.

Things have to be maintained and new feature requests will fly in once a command is there while time is limited.  Weighing the cost-benefit ratio is challenging.

I generally prefer keeping things small.  But that does not mean the feature won't happen :)

**Comment by rhatdan** [comment]
@cdoern any update on this one?

**Comment by cdoern** [comment]
@rhatdan what should we do about this?

**Comment by rhatdan** [comment]
I thought you were working on this?

**Comment by cdoern** [comment]
Didn't we decide that we would table this until `generate spec` merged? Cc/ @baude

**Comment by rhatdan** [comment]
Yes, but this is part of that effort.
